### PR TITLE
Vagrant Script: Detect colliding active virtualbox VMs and warn users

### LINKED ISSUE #13928: vagrant script: detect colliding active virtualbox VMs and warn users
Running the vagrant script with `NWORKERS=1` did not work for me. Specifically, `k8s2` could not connect to the API server even though everything (e.g., certificates) looked fine on the host. After a long debugging session, we figure out that the issue was a residual `k8s1` machine that was running on virtualbox but was not visible to vagrant.

It would be good to detect these types of residual VMs and warn the users when they run the script.

### PR DESCRIPTION
**Pull request by ghost**
By running vagrant script when there is active residual VMs that are not
visible to vagrant creates error.

This commit fix that error by listing the active residual VMs and
presenting the user with options to Pause, Stop or Destroy the residual
VMs or ignore and continue.

...
$NO_BUILD=1 K8S=1 ./contrib/vagrant/start.sh

Detected running VMs that might cause conflict

"cilium_k8s1_1610031382439_78313" {a3c160c3-9f2b-4aae-8c87-44e004a7a003}

Do you wish to pause, stop, or destroy the VM(s) or ignore and continue?
[p/s/d/C]
...

Fixes: #13928

Signed-off-by: Vishnu Soman K <vishnusomank05@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #issue-number

### CODE REVIEW

**Comment by vsk-coding** [comment]
@pchaigno Sorry I accidently deleted that branch. I am very new to contribution. Can you take a look at the code?

File: contrib/vagrant/start.sh [hunk 1]
```diff
@@ -576,36 +576,110 @@ function vboxnet_addr_finder(){
...
+	echo "Detected running VMs that might cause conflict"
+	echo
+	VBoxManage list runningvms
+	echo
+	printf "Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?  "
+	#VBoxManage list runningvms
+    VBoxManage list runningvms
+	#option to pause, stop or delete VM(s) or ignore and continue
+    echo
+	printf "Do you wish to pause, stop, or destroy the VM(s) or ignore and continue? [p/s/d/C]  "
```
Comment (pchaigno) [inline contrib/vagrant/start.sh]: This message is now printed twice (here and line 623 above).

File: contrib/vagrant/start.sh [hunk 2]
```diff
@@ -576,36 +576,110 @@ function vboxnet_addr_finder(){
...
+				echo
+				echo $VMName
+				VBoxManage controlvm $VMName poweroff
+				VBoxManage unregistervm $VMName --delete				
+				printf "\n$VMName destroyed\n"				
+			done
+			printf "\n$runningVm VM(s) destroyed successfully\n"
+			printf "Tip: Try running vagrant status to check status\n\n"
+		;;
+		"C" )
```
Comment (pchaigno) [inline contrib/vagrant/start.sh]: We need to make this the default case.

File: contrib/vagrant/start.sh [hunk 3]
```diff
@@ -576,36 +576,110 @@ function vboxnet_addr_finder(){
...
+		"p" )
+			for ((i=1; i<=$runningVm; i=i+1))
+			do
+                #checks state of the VM(s) before trying to pause as paused VM(s) are listed in runningvms command
+				VMName=$(VBoxManage list runningvms | awk 'NR=='$i'{print $1}' |  cut -d "\"" -f 2)
+				echo
+				echo $VMName
+				VBoxManage controlvm $VMName pause
+				if [ $(VBoxManage showvminfo $VMName | grep -c "paused (since") -eq 0 ]; then
+                    VBoxManage controlvm $VMName pause
```
Comment (pchaigno) [inline contrib/vagrant/start.sh]: Why are we trying to pause the VM twice?

File: contrib/vagrant/start.sh [hunk 4]
```diff
@@ -576,36 +576,110 @@ function vboxnet_addr_finder(){
...
+					printf "\n$VMName paused\n"
+				else
+					printf "\n$VMName already in paused state\n"
+					temp=$((temp-1))
+				fi
+			done
+			printf "\n$runningVm VM(s) paused successfully\n" 
+			printf "\n$temp VM(s) paused successfully\n"
```
Comment (pchaigno) [inline contrib/vagrant/start.sh]: Looks like a duplicated message.

File: contrib/vagrant/start.sh [hunk 5]
```diff
@@ -576,36 +576,110 @@ function vboxnet_addr_finder(){
...
+	printf "Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?  "
+	#VBoxManage list runningvms
+    VBoxManage list runningvms
+	#option to pause, stop or delete VM(s) or ignore and continue
+    echo
+	printf "Do you wish to pause, stop, or destroy the VM(s) or ignore and continue? [p/s/d/C]  "
+	read optn
+    temp=$runningVm
+	case "$optn" in
+		"p" )
```
Comment (pchaigno) [inline contrib/vagrant/start.sh]: I still don't think `pause` is a viable option. If we choose to pause the VM on the prompt, then we still get the prompt again on the next `start.sh` run.

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: contrib/vagrant/start.sh (hunk 5 above)
Comment (ghost) [reply contrib/vagrant/start.sh]: removed the pause option

**Comment by vsk-coding** [comment]
test-me-please

**Comment by pchaigno** [comment]
> test-me-please

Only members of @cilium can run end-to-end tests. They are not required for this pull request though.

It looks like the current version breaks `NWORKERS=1 ./contrib/vagrant/start.s [second_vm_name]`:

```
$ NO_BUILD=1 K8S=1 NETNEXT=1 NWORKERS=1 ./contrib/vagrant/start.sh k8s2+

Detected running VMs that might cause conflict

"cilium_k8s1_1610970336448_75825" {6994c5a7-2507-4d99-af62-a2e54e5eb5fd}

Do you wish to stop, destroy the VM(s) or ignore and continue? [s/d/C]  

# NFS enabled. don't forget to enable these ports on your host
# before starting the VMs in order to have nfs working
# iptables -I INPUT -p tcp -s 192.168.34.0/24 --dport 111 -j ACCEPT
# iptables -I INPUT -p tcp -s 192.168.34.0/24 --dport 2049 -j ACCEPT
# iptables -I INPUT -p tcp -s 192.168.34.0/24 --dport 20048 -j ACCEPT
# To use kubectl on the host, you need to add the following route:
# ip route add 192.168.33.11 via 192.168.34.11
Bringing machine 'k8s1+' up with 'virtualbox' provider...
Bringing machine 'k8s2+' up with 'virtualbox' provider...
==> k8s1+: Machine already provisioned. Run `vagrant provision` or use the `--provision`
==> k8s1+: flag to force provisioning. Provisioners marked to run always will still run.
[...]
==> k8s1+: Running provisioner: ipv6-config (shell)...
    k8s1+: Running: inline script
    k8s1+: RTNETLINK answers: File exists
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

It looks like it just ignores the argument to `start.sh`.

**Comment by vsk-coding** [comment]
@pchaigno i've updated the code to include the missing flow. can you please take a look at it.

File: contrib/vagrant/start.sh [hunk 6]
```diff
@@ -576,36 +576,91 @@ function vboxnet_addr_finder(){
...
+runningVm=$(VBoxManage list runningvms | awk 'END{ print NR }')
+VMName=$(VBoxManage list runningvms | awk 'NR==1{print $1}' |  cut -d "\"" -f 2)
+if [ "$VMName" ]; then
+    echo
+	echo "Detected running VMs that might cause conflict"
+	echo
+    VBoxManage list runningvms
+
+	#option to stop, delete VM(s) or ignore and continue
+    echo
```
Comment (pchaigno) [inline contrib/vagrant/start.sh:625]: Could you fix up all indentations in this file? It looks like there's a mix of tabulations and spaces.

File: contrib/vagrant/start.sh [hunk 7]
```diff
@@ -576,36 +576,91 @@ function vboxnet_addr_finder(){
...
+            echo
+            #value to start.sh is passed to function createVm
+			createVm $1
+		;;
+	esac
+else
+    echo
+    #value to start.sh is passed to function createVm
+	createVm $1
+fi
```
Comment (pchaigno) [inline contrib/vagrant/start.sh]: Missing newline here.

**Review by pchaigno — COMMENTED** [review]
It looks great! Thanks @vsk-coding!

Only two minor code style issues to fix below. I think we'll be good after that.

File: contrib/vagrant/start.sh (hunk 7 above)
Comment (ghost) [reply contrib/vagrant/start.sh]: added the missing newline and fixed style issues

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
@aditighag Any chance you could give a try to this on macOS? It should be enough to try to provision a VM with one already started (e.g., k8s1 vs. k8s2 or k8s1+ vs. runtime1); it should print a warning and provide several choices. I mostly want to make sure we're not breaking anything on other OSes.

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR! Could you squash the two commits into one?

**Comment by vsk-coding** [comment]
@christarazi I've squashed the commits as asked.

**Comment by aditighag** [comment]
It works for me on mac OS. Although I'm not sure why it prompts about the test VMs? Dev and test VMs are provisioned independently so they shouldn't "collide" with each other?

>NO_BUILD=1 NWORKERS=0 NETNEXT=1 K8S=1 ./contrib/vagrant/start.sh k8s1+

Detected running VMs that might cause conflict

"cilium_k8s1_1605748387242_13783" {2cef177e-b01e-4975-ae88-878a98931b9d}
"test_k8s1-120_1611817099500_97283" {5dcec1cf-ca0e-4cee-9404-742e1ae552d6}
"test_k8s2-120_1611817760993_69394" {a9605716-9db3-4b55-9644-b6378f4d45d9}

Do you wish to stop, destroy the VM(s) or ignore and continue? [s/d/C]

**Comment by pchaigno** [comment]
> Dev and test VMs are provisioned independently so they shouldn't "collide" with each other?

Are you sure they can't? The initial issue here was a conflict between IP addresses and I suspect the same can happen between test and dev. VMs.

In any case, I think it's fine to merge as is and address that case as a follow up.

**Comment by pchaigno** [comment]
@vsk-coding You'll need to rebase to fix the failing GitHub action. We had a regression a couple days ago and fixed it in master.

**Comment by pchaigno** [comment]
test-me-please

**Comment by vsk-coding** [comment]
> @vsk-coding You'll need to rebase to fix the failing GitHub action. We had a regression a couple days ago and fixed it in master.

I've rebased and pushed the code. is there anything more to do from my end?

**Comment by pchaigno** [comment]
I've confirmed that the latest version works and run extensive tests on prior versions. Aditi confirmed it works on macOS. I'm marking as ready to merge, we can rework the test vs. dev. VM detection later if needed.

**Review by tklauser — APPROVED** [review]

**Comment by pchaigno** [comment]
@vsk-coding Thanks a lot for tackling this! :tada: :rocket:

**Comment by vsk-coding** [comment]
> @vsk-coding Thanks a lot for tackling this! 🎉 🚀

Thanks for the guidance and patience 🌟👍

**Pull request merged**